### PR TITLE
Rc realtime dev

### PR DESCRIPTION
**Pull request by qdpham13**
- Refactor 
    - ConfigRealtimeHttpClient and ConfigAutoFetch no longer need to share the HttpUrlConnection object; it is completely local to ConfigAutofetch and doesn't need to be concurrent safe.
    - Turn ConfigAutofetch into a runnable and execute it from ConfigRealtimeHttpClient using ExecutorService and keep track of future. 
- Add retrying methods so that if a HttpUrlConnection can't be made, it retries. Includes exponential backoff and retry limit.
- Add background listener, taken from FirebaseApp (https://github.com/firebase/firebase-android-sdk/blob/87ae78e9c149a9794b5956682e7c61afb27a628c/firebase-common/src/main/java/com/google/firebase/FirebaseApp.java#L672) so that we can retry everytime the app goes from background to foreground.
- Expose public realtime api.

### CODE REVIEW

**Comment by qdpham13** [comment]
PTAL when you get the chance, thanks! @rlazo @vkryachko

File: firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/TestConstructorUtil.kt [hunk 1]
```diff
@@ -50,6 +52,7 @@ fun createRemoteConfig(
             defaultConfigsCache,
             fetchHandler,
             getHandler,
-            frcMetadata
+            frcMetadata,
+                realtimeClient
```
Comment (rlazo) [inline firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/TestConstructorUtil.kt:56]: indentation

File: firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/ktx/RemoteConfigTests.kt [hunk 2]
```diff
@@ -137,7 +138,9 @@ class ConfigTests : BaseTestCase() {
             defaultConfigsCache = mock(ConfigCacheClient::class.java),
             fetchHandler = mock(ConfigFetchHandler::class.java),
             getHandler = mockGetHandler,
-            frcMetadata = mock(ConfigMetadataClient::class.java))
+            frcMetadata = mock(ConfigMetadataClient::class.java),
+                realtimeClient = mock(ConfigRealtimeHttpClient::class.java)
```
Comment (rlazo) [inline firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/ktx/RemoteConfigTests.kt:142]: indentation

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/FirebaseRemoteConfig.java [hunk 3]
```diff
@@ -541,6 +545,19 @@ public Task<Void> reset() {
         });
   }
 
+  /**
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/FirebaseRemoteConfig.java:548]: Maybe developers have more context, but from just reading the comment, I'm not sure what "start realtime and return registration" means

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java [hunk 4]
```diff
@@ -199,9 +207,11 @@ synchronized FirebaseRemoteConfig get(
               defaultsClient,
               fetchHandler,
               getHandler,
-              metadataClient);
+              metadataClient,
+              getRealtime(firebaseApp, firebaseInstallations, fetchHandler, context, namespace));
       in.startLoadingConfigsFromDisk();
       frcNamespaceInstances.put(namespace, in);
+      frcNamespaceInstancesBackground.put(namespace, in);
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java:214]: All instances are both in `frcNamespaceInstances` and `frcNamespaceInstancesBackground`. Why do you need both maps?

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java [hunk 5]
```diff
@@ -299,4 +324,43 @@ private static boolean isAbtSupported(FirebaseApp firebaseApp, String namespace)
...
     return firebaseApp.getName().equals(FirebaseApp.DEFAULT_APP_NAME);
   }
+
+  private synchronized void notifyRCInstances(boolean background) {
+    for (FirebaseRemoteConfig frc : frcNamespaceInstancesBackground.values()) {
+      if (!background) {
+        // Add dummy listener and remove to trigger http stream flow.
+        ConfigUpdateListenerRegistration registration =
+            frc.addOnConfigUpdateListener(
+                new ConfigUpdateListener() {
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java:334]: Nit. You can optionally create a constant named `EMPTY_CONFIG_LISTENER` or `NOOP_CONFIG_LISTENER`, set it to be this object, and assign it here. That would make this snippet easier to read.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java [hunk 6]
```diff
@@ -299,4 +324,43 @@ private static boolean isAbtSupported(FirebaseApp firebaseApp, String namespace)
...
+        GlobalBackgroundListener globalBackgroundListener = new GlobalBackgroundListener();
+        if (INSTANCE.compareAndSet(null, globalBackgroundListener)) {
+          BackgroundDetector.initialize(application);
+          BackgroundDetector.getInstance().addListener(globalBackgroundListener);
+        }
+      }
+    }
+
+    @Override
+    public void onBackgroundStateChanged(boolean b) {
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java:362]: `b` is nondescriptive.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java [hunk 7]
```diff
@@ -32,52 +32,42 @@
...
-      ConfigUpdateListener retryCallback,
-      Executor executor,
-      ScheduledExecutorService scheduledExecutorService) {
+      ConfigUpdateListener retryCallback) {
     this.httpURLConnection = httpURLConnection;
     this.configFetchHandler = configFetchHandler;
     this.eventListeners = eventListeners;
     this.retryCallback = retryCallback;
-    this.scheduledExecutorService = scheduledExecutorService;
+    this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java:64]: It's usually a good idea to receive the executor as a parameter to make testing easier, what's the rationale behind the change?

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java [hunk 8]
```diff
@@ -121,19 +119,26 @@ private void handleNotifications(InputStream inputStream) throws IOException {
...
     String message;
     while ((message = reader.readLine()) != null) {
+      if (!message.contains("latestTemplateVersionNumber")) {
+        continue;
+      }
+
       long targetTemplateVersion = configFetchHandler.getTemplateVersionNumber();
       try {
-        JSONObject jsonObject = new JSONObject(message);
+        JSONObject jsonObject = new JSONObject("{" + message + "}");
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java:128]: Could it happen that message does include the `{}`?

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 9]
```diff
@@ -62,34 +67,49 @@ public class ConfigRealtimeHttpClient {
...
-  private HttpURLConnection httpURLConnection;
+  private Future<?> autoFetchTask;
+
+  @GuardedBy("this")
+  private int RETRIES_REMAINING;
+
+  @GuardedBy("this")
+  private long RETRY_TIME_SECONDS;
+
+  @GuardedBy("this")
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:78]: Instances of `Random` are thread safe https://developer.android.com/reference/java/util/Random

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java [hunk 10]
```diff
@@ -62,34 +67,49 @@ public class ConfigRealtimeHttpClient {
   private final Set<ConfigUpdateListener> listeners;
 
   @GuardedBy("this")
-  private HttpURLConnection httpURLConnection;
+  private Future<?> autoFetchTask;
+
+  @GuardedBy("this")
+  private int RETRIES_REMAINING;
```
Comment (rlazo) [inline firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java:73]: Uppercase is used only for constant fields.

File: firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/TestConstructorUtil.kt (hunk 1 above)
Comment (qdpham13) [reply firebase-config/ktx/src/test/kotlin/com/google/firebase/remoteconfig/TestConstructorUtil.kt]: Fixed.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/FirebaseRemoteConfig.java (hunk 3 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/FirebaseRemoteConfig.java:548]: Updated to be similar to Firestore's
https://github.com/firebase/firebase-android-sdk/blob/cf4bcfd23586e39ab9e127aeaa30bc39a9a7da66/firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java#L361

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java (hunk 4 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java]: Originally the background listener was static so I had a static map. Changed to use frcNamespaceInstances.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java (hunk 6 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/RemoteConfigComponent.java]: updated to a more descriptive variable name. b to isInBackground

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java (hunk 7 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java:64]: The idea was so that each autoFetch thread would get an executor to handle fetches. That way when the autoFetch thread dies/ends we can also stop the respective scheduled executor to close the fetch threads and so they don't stay running.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java (hunk 8 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java:128]: No, since we're using chunked encoding every line of the response is sent separately. So that's why we need to add the curly brackets; the brackets will be sent a the beginning and end of every full message.

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java (hunk 9 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java]: Removed guardedby

File: firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java (hunk 10 above)
Comment (qdpham13) [reply firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java]: Updated to normal camel case spelling.

**Comment by qdpham13** [comment]
> PTAL when you get the chance, thanks! @rlazo @vkryachko

Bumping this up to get more visibility, hoping to get this reviewed/approved by the end of the week and before holidays next week. Thanks! @rlazo @vkryachko
